### ISSUE TITLE
Make the provider instance filterable

### ISSUE THREAD

**Issue by tfrommen** [humanmade/asset-manager-framework#15]
Currently, there is a filter for the provider **class** to use, and then AMF will take care of instantiating.

If AMF provided a filter for the provider **instance**, this would allow for both decorating providers (see example below) and constructor injection in general.

This filter would ideally replace the current one, otherwise the provider constructor cannot have mandatory arguments.

----

**Current code:**

```php
function get_provider() : Provider {
	static $provider = null;

	if ( $provider ) {
		return $provider;
	}

	$provider_class = apply_filters( 'amf/provider_class', __NAMESPACE__ . '\BlankProvider' );
	$provider = new $provider_class();

	return $provider;
}
```

**Proposed code:**

```diff
function get_provider() : Provider {
	static $provider = null;

	if ( $provider ) {
		return $provider;
	}

-	$provider_class = apply_filters( 'amf/provider_class', __NAMESPACE__ . '\BlankProvider' );
-	$provider = new $provider_class();
+	$provider = apply_filters( 'amf/provider', new BlankProvider() );

	return $provider;
}
```

**Simple usage:**

```php
add_filter( 'amf/provider', function () {
    return new CustomProvider( 'some', 'stuff' );
} );
```

**Decorator usage:**

```php
add_filter( 'amf/provider', function ( Provider $provider ) {
    return new DecoratingProvider( $provider, 'some', 'stuff' );
} );
```

The `DecoratingProvider` could then store the current `$provider` instance, and delegate to it as necessary, or perform custom logic.

**Comment by johnbillion** [comment]
Agreed.

**Comment by johnbillion** [comment]
Fixed in #16
